### PR TITLE
Sandbox sizing feature

### PR DESCRIPTION
**Pull request by egernst**
There are software and hardware architectures which do not support dynamically adjusting the CPU and memory resources associated with a sandbox. For these, today, they rely on "default CPU" and "default memory" configuration options for the runtime, either set by annotation or by the configuration toml on disk.

In the case of a single container (launched by ctr, or something like "docker run"), we could allow for sizing the VM correctly, since all of the information is already available to us at creation time.

In the sandbox / pod container case, it is possible for the upper layer container runtime (ie, containerd or crio) could send a specific annotation indicating the total workload resource requirements associated with the sandbox creation request. 

In either case, we should start keeping track of the sandbox resources we want, and the base amount of CPU/memory provided to the VM. In the instance of not sizing information being provided, we will follow same behavior as today: start the VM with the default CPU/memory.

If information *is* provided, we should start the VM with the default CPU/memory + the workload CPU/memory that was specified.

In this scenario (we start the VM with the "right" amount of CPU/Memory), we do not want to update the VM resources when containers are added, or adjusted in size.

This functionality is introduced behind a configuration flag, `enable_static_resource_management`. This'll greatly improve UX for folks who are utilizing Kata with a VMM or hardware architecture that doesn't support hotplug.

Note, users will still be unable to do in place vertical pod autoscaling or other dynamic container/pod sizing with this enabled.

### To do:
 - [x] add toml/config plumbing for `enable_static_resource_management` flag
 - [x] add unit tests
 - [x] docs

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @egernst - Please could you ensure that when you raise PRs, there all have an explanatory comment and a clear reference to an issue?

I've seen a few other "bare" PRs like this recently. Without the standard summary information, it's much harder for the community to review, which means the time to land such PRs is going to increase significantly imho.

If these sorts of PRs are draft / rfc / experiments, please let us know by adding tags to the title, or using the appropriate labels.

Thanks for your understanding! :smile:

**Comment by egernst** [comment]
Sorry about that; first start for abusing the CI; I need to have a better habit for the draft PR! Thanks James.

**Comment by jodh-intel** [comment]
Marking as stale as this PR may have been abandoned.

**Comment by egernst** [comment]
/test

un-stale-ing

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (fengwang666) [inline src/runtime/pkg/containerd-shim-v2/create.go:66]: nit: use `uint64`?

File: src/runtime/virtcontainers/sandbox.go
Comment (fengwang666) [inline src/runtime/virtcontainers/sandbox.go:107]: `requiredf or` -> `required for`

File: src/runtime/pkg/oci/utils.go
Comment (fengwang666) [inline src/runtime/pkg/oci/utils.go:935]: nit: `workload mem` -> `workload mem in MB`?

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (egernst) [reply src/runtime/pkg/containerd-shim-v2/create.go:66]: This is done to match what is defined in LinuxResources. Why memory is int64 instead of unit is ... :shrug:

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (egernst) [inline src/runtime/pkg/containerd-shim-v2/create.go:53]: While these aren't used anywhere else, I wonder if these functions would be better suited in a utility package. Having said that, I see that there are about 4 utility packages in the repo today, and which would be best isn't yet clear. I'm happy to take this on in follow up, or here if folks have opinions.

File: src/runtime/config/configuration-clh.toml.in
Comment (sameo) [inline src/runtime/config/configuration-clh.toml.in:189]: You may want to move that into a separate commit/PR?

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (sameo) [reply src/runtime/pkg/containerd-shim-v2/create.go:53]: Would it make sense to have them both take a `*specs.Spec` as an input and move them to `vc/pkg/oci`?

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (sameo) [inline src/runtime/pkg/containerd-shim-v2/create.go:81]: What happens when resources is nil? I guess this is returning `(0,0)`?

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (egernst) [reply src/runtime/pkg/containerd-shim-v2/create.go:53]: Yes, agreed.

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (egernst) [reply src/runtime/pkg/containerd-shim-v2/create.go:81]: Yes, that's what it'll do. 

Let me do a quick return at top of function for this scenario - it'll make it more clear, and skip some code in the likely scenario.

File: src/runtime/config/configuration-clh.toml.in
Comment (egernst) [reply src/runtime/config/configuration-clh.toml.in:220]: done!

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (egernst) [reply src/runtime/pkg/containerd-shim-v2/create.go]: Done!

File: src/runtime/pkg/containerd-shim-v2/create.go
Comment (bergwolf) [inline src/runtime/pkg/containerd-shim-v2/create.go:115]: What if there is no upper layer annotation for sandbox sizing, and `static_sandbox_resource_mgmt` is enabled?

Comment (egernst) [reply src/runtime/pkg/containerd-shim-v2/create.go:115]: For SW/HW architectures w/ out hotplug, this'll behave the same. The VM will be sized based on default CPU/memory from the toml, and won't be adjusted after boot.

File: src/runtime/pkg/oci/utils.go
Comment (fidencio) [inline src/runtime/pkg/oci/utils.go:1095]: Hmm. Please, give me some log about this when this happens.  We don't need to fail, but info log, which would only be displayed when we enable debug, would be enough to help debugging possible issues with this.

**Review by fidencio — CHANGES_REQUESTED** [review]
@egernst, in general the PR looks good.
I've made a few suggestions so we can ensure, by documentation, that folks won't try to use it together with a combination of Kubernetes / CRI Engine that does not provide such functionality yet.

Last but not least, there's one place that I could found where some logs would be helpful for debugging purposes in the future.

File: docs/design/vcpu-handling.md
Comment (c3d) [inline docs/design/vcpu-handling.md:165]: Why do you need a flag, if we can statically determine it based on hypervisor?
* For users of the hypervisor that need it, that means they need to enable it, so it's inconvenient
* Is there any valid reason to enable it e.g. for qemu?

File: src/runtime/config/configuration-clh.toml.in
Comment (egernst) [reply src/runtime/config/configuration-clh.toml.in:236]: I can add that for the kubernetes case, but this will also be relevant in the non-kubernetes scenario (when launching with a tool _like_ ctr)

File: src/runtime/pkg/oci/utils.go
Comment (egernst) [reply src/runtime/pkg/oci/utils.go:1095]: This is actually the "normal" scenario, where the annotation is not provided. I think a debug log for what the retrieved values are, however (so you can see we are setting as 0 versus a set value?)

File: docs/design/vcpu-handling.md
Comment (egernst) [reply docs/design/vcpu-handling.md:177]: In the case of Kubernetes, today a "pause container" is created with the sandbox start request. It is this "infrastructure pods" container's config.json.

File: docs/design/vcpu-handling.md
Comment (egernst) [reply docs/design/vcpu-handling.md:160]: Through CRI / shim-v2, we are only received quota, period and shares.

File: src/runtime/pkg/oci/utils.go
Comment (egernst) [reply src/runtime/pkg/oci/utils.go:1095]: Actually, let me just expand each parse in order to check and log a warning. I was excited to be concise, but that isn't necessarily best here!

Comment (fidencio) [reply src/runtime/pkg/oci/utils.go:1095]: Thanks, @egernst.

File: src/runtime/config/configuration-clh.toml.in
Comment (fidencio) [reply src/runtime/config/configuration-clh.toml.in:236]: Yep, add a note about `ctr` as usual, if possible.

What I'm trying to avoid here is a bunch of issues opened because someone tries to use something and has absolutely no information on whether that should work or not with the setup they have (eg: k8s, containerd, cri-o, ctr, ...).   As the matrix of elements is not **that** huge (only k8s, containerd, cri-o, and ctr) I think it's worth mentioning.

Sounds reasonable?

File: src/runtime/pkg/oci/utils.go
Comment (jodh-intel) [inline src/runtime/pkg/oci/utils.go:1098]: Can you add a reference that shows the expected types of these three annotations?

File: src/runtime/virtcontainers/sandbox.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/sandbox.go:1929]: Nit: `Debug()` is sufficient as there are not formats in the string.

File: docs/design/vcpu-handling.md
Comment (jodh-intel) [inline docs/design/vcpu-handling.md:163]: You could link to https://github.com/kata-containers/kata-containers/blob/main/docs/hypervisors.md here maybe?

**Review by jodh-intel — COMMENTED** [review]
Thanks @egernst - a few comments...

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

File: src/runtime/pkg/oci/utils.go
Comment (egernst) [reply src/runtime/pkg/oci/utils.go:1109]: I wanted to have a consistent parameter between the two sizing calls. Perhaps that isn't that important, though....

File: src/runtime/config/configuration-clh.toml.in
Comment (egernst) [reply src/runtime/config/configuration-clh.toml.in:236]: updated, PTAL.

File: docs/design/vcpu-handling.md
Comment (egernst) [reply docs/design/vcpu-handling.md:163]: Since we don't specifically call out CPU / memory hotplug in the hypervisor document, I'm not sure this would be super helpful here.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @egernst!

**Review by sameo — APPROVED** [review]

**Review by c3d — APPROVED** [review]
Looks good and useful to me. Thanks @egernst

**Pull request merged**
